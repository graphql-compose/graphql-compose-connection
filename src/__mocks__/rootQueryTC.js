/* @flow */

import { schemaComposer } from 'graphql-compose';

export const rootQueryTC = schemaComposer.createObjectTC({
  name: 'RootQuery',
  fields: {},
});
