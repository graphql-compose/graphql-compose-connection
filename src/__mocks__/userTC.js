/* @flow */
/* eslint-disable no-param-reassign */

import { schemaComposer } from 'graphql-compose';
import {
  GraphQLString,
  GraphQLObjectType,
  GraphQLInputObjectType,
  GraphQLEnumType,
  GraphQLInt,
} from 'graphql-compose/lib/graphql';
import type { ConnectionSortMapOpts } from '../connectionResolver';

export const UserType = new GraphQLObjectType({
  name: 'User',
  fields: {
    id: {
      type: GraphQLInt,
    },
    name: {
      type: GraphQLString,
    },
    age: {
      type: GraphQLInt,
    },
    gender: {
      type: GraphQLString,
    },
  },
});

export const userTC = schemaComposer.createObjectTC(UserType);

export const userList = [
  { id: 1, name: 'user01', age: 11, gender: 'm' },
  { id: 2, name: 'user02', age: 12, gender: 'm' },
  { id: 3, name: 'user03', age: 13, gender: 'f' },
  { id: 4, name: 'user04', age: 14, gender: 'm' },
  { id: 5, name: 'user05', age: 15, gender: 'f' },
  { id: 6, name: 'user06', age: 16, gender: 'f' },
  { id: 7, name: 'user07', age: 17, gender: 'f' },
  { id: 8, name: 'user08', age: 18, gender: 'm' },
  { id: 9, name: 'user09', age: 19, gender: 'm' },
  { id: 10, name: 'user10', age: 49, gender: 'f' },
  { id: 11, name: 'user11', age: 49, gender: 'm' },
  { id: 12, name: 'user12', age: 47, gender: 'f' },
  { id: 15, name: 'user15', age: 45, gender: 'm' },
  { id: 14, name: 'user14', age: 45, gender: 'm' },
  { id: 13, name: 'user13', age: 45, gender: 'f' },
];

const filterArgConfig = {
  name: 'filter',
  type: new GraphQLInputObjectType({
    name: 'FilterUserInput',
    fields: {
      gender: {
        type: GraphQLString,
      },
      age: {
        type: GraphQLInt,
      },
    },
  }),
};

function filteredUserList(list, filter = {}) {
  let result = list.slice();
  if (filter.gender) {
    result = result.filter(o => o.gender === filter.gender);
  }

  if (filter.id) {
    if (filter.id.$lt) {
      result = result.filter(o => o.id < filter.id.$lt);
    }
    if (filter.id.$gt) {
      result = result.filter(o => o.id > filter.id.$gt);
    }
  }
  if (filter.age) {
    if (filter.age.$lt) {
      result = result.filter(o => o.age < filter.age.$lt);
    }
    if (filter.age.$gt) {
      result = result.filter(o => o.age > filter.age.$gt);
    }
  }

  return result;
}

function sortUserList(list, sortValue = {}) {
  const fields = Object.keys(sortValue);
  list.sort((a, b) => {
    let result = 0;
    fields.forEach(field => {
      if (result === 0) {
        if (a[field] < b[field]) {
          result = sortValue[field] * -1;
        } else if (a[field] > b[field]) {
          result = sortValue[field];
        }
      }
    });
    return result;
  });
  return list;
}

function prepareFilterFromArgs(resolveParams = {}) {
  const args = resolveParams.args || {};
  const filter = Object.assign({}, args.filter);
  if (resolveParams.rawQuery) {
    Object.keys(resolveParams.rawQuery).forEach(k => {
      filter[k] = resolveParams.rawQuery[k];
    });
  }
  return filter;
}

export const findManyResolver = schemaComposer.createResolver({
  name: 'findMany',
  kind: 'query',
  type: UserType,
  args: {
    filter: filterArgConfig,
    sort: new GraphQLEnumType({
      name: 'SortUserInput',
      values: {
        ID_ASC: { value: { id: 1 } },
        ID_DESC: { value: { id: -1 } },
        AGE_ASC: { value: { age: 1 } },
        AGE_DESC: { value: { age: -1 } },
      },
    }),
    limit: GraphQLInt,
    skip: GraphQLInt,
  },
  resolve: resolveParams => {
    const args = resolveParams.args || {};
    const { sort, limit, skip } = args;

    let list = userList.slice();
    list = sortUserList(list, sort);
    list = filteredUserList(list, prepareFilterFromArgs(resolveParams));

    if (skip) {
      list = list.slice(skip);
    }

    if (limit) {
      list = list.slice(0, limit);
    }

    return Promise.resolve(list);
  },
});
userTC.setResolver('findMany', findManyResolver);

export const countResolver = schemaComposer.createResolver({
  name: 'count',
  kind: 'query',
  type: GraphQLInt,
  args: {
    filter: filterArgConfig,
  },
  resolve: resolveParams => {
    return Promise.resolve(filteredUserList(userList, prepareFilterFromArgs(resolveParams)).length);
  },
});
userTC.setResolver('count', countResolver);

export const sortOptions: ConnectionSortMapOpts = {
  ID_ASC: {
    value: { id: 1 },
    cursorFields: ['id'],
    beforeCursorQuery: (rawQuery, cursorData) => {
      if (!rawQuery.id) rawQuery.id = {};
      rawQuery.id.$lt = cursorData.id;
    },
    afterCursorQuery: (rawQuery, cursorData) => {
      if (!rawQuery.id) rawQuery.id = {};
      rawQuery.id.$gt = cursorData.id;
    },
  },
  AGE_ID_DESC: {
    value: { age: -1, id: -1 },
    cursorFields: ['age', 'id'],
    beforeCursorQuery: (rawQuery, cursorData) => {
      if (!rawQuery.age) rawQuery.age = {};
      if (!rawQuery.id) rawQuery.id = {};
      rawQuery.age = { $gt: cursorData.age };
      rawQuery.id = { $gt: cursorData.id };
    },
    afterCursorQuery: (rawQuery, cursorData) => {
      if (!rawQuery.age) rawQuery.age = {};
      if (!rawQuery.id) rawQuery.id = {};
      rawQuery.age = { $lt: cursorData.age };
      rawQuery.id = { $lt: cursorData.id };
    },
  },
};
