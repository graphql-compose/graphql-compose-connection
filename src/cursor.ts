export type CursorDataType =
  | {
      [fieldName: string]: any;
    }
  | any;

export function base64(i: string): string {
  return Buffer.from(i, 'ascii').toString('base64');
}

export function unbase64(i: string): string {
  return Buffer.from(i, 'base64').toString('ascii');
}

export function cursorToData(cursor?: string | any): CursorDataType | number | null {
  if (typeof cursor === 'string') {
    try {
      return JSON.parse(unbase64(cursor)) || null;
    } catch (err) {
      return null;
    }
  }
  return null;
}

export function dataToCursor(data: CursorDataType | number): string {
  return base64(JSON.stringify(data));
}
