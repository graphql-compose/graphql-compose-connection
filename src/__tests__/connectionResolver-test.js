/* @flow */
/* eslint-disable no-param-reassign */

import { Resolver } from 'graphql-compose';
import { GraphQLInt, GraphQLString } from 'graphql-compose/lib/graphql';
import { userTC, userList, sortOptions } from '../__mocks__/userTC';
import { dataToCursor } from '../cursor';
import { prepareConnectionResolver, prepareRawQuery, preparePageInfo } from '../connectionResolver';

describe('connectionResolver', () => {
  const connectionResolver = prepareConnectionResolver(userTC, {
    countResolverName: 'count',
    findResolverName: 'findMany',
    sort: sortOptions,
  });

  describe('definition checks', () => {
    it('should return Resolver', () => {
      expect(connectionResolver).toBeInstanceOf(Resolver);
    });

    it('should throw error if first arg is not ObjectTypeComposer', () => {
      expect(() => {
        const wrongArgs: any = [123];
        prepareConnectionResolver(...wrongArgs);
      }).toThrowError('should be instance of ObjectTypeComposer');
    });

    it('should throw error if opts.countResolverName are empty', () => {
      expect(() => {
        const wrongArgs: any = [userTC, {}];
        prepareConnectionResolver(...wrongArgs);
      }).toThrowError('should have option `opts.countResolverName`');
    });

    it('should throw error if resolver opts.countResolverName does not exists', () => {
      expect(() =>
        prepareConnectionResolver(userTC, {
          countResolverName: 'countDoesNotExists',
          findResolverName: 'findMany',
          sort: sortOptions,
        })
      ).toThrowError("does not have resolver with name 'countDoesNotExists'");
    });

    it('should throw error if opts.findResolverName are empty', () => {
      expect(() => {
        const wrongArgs: any = [userTC, { countResolverName: 'count' }];
        prepareConnectionResolver(...wrongArgs);
      }).toThrowError('should have option `opts.findResolverName`');
    });

    it('should throw error if resolver opts.countResolverName does not exists', () => {
      expect(() => {
        const wrongArgs: any = [
          userTC,
          {
            countResolverName: 'count',
            findResolverName: 'findManyDoesNotExists',
            sort: sortOptions,
          },
        ];
        prepareConnectionResolver(...wrongArgs);
      }).toThrowError("does not have resolver with name 'findManyDoesNotExists'");
    });
  });

  describe('resolver basic properties', () => {
    it('should have name `connection`', () => {
      expect(connectionResolver.name).toBe('connection');
    });

    it('should have kind `query`', () => {
      expect(connectionResolver.kind).toBe('query');
    });

    it('should have type to be ConnectionType', () => {
      expect((connectionResolver.type: any).name).toBe('UserConnection');
    });
  });

  describe('resolver args', () => {
    it('should have `first` arg', () => {
      expect(connectionResolver.getArgType('first')).toBe(GraphQLInt);
    });

    it('should have `last` arg', () => {
      expect(connectionResolver.getArgType('last')).toBe(GraphQLInt);
    });

    it('should have `after` arg', () => {
      expect(connectionResolver.getArgType('after')).toBe(GraphQLString);
    });

    it('should have `before` arg', () => {
      expect(connectionResolver.getArgType('before')).toBe(GraphQLString);
    });

    it('should have `sort` arg', () => {
      expect((connectionResolver.getArgType('sort'): any).name).toBe('SortConnectionUserEnum');
    });
  });

  describe('call of resolvers', () => {
    let spyResolveParams;
    let mockedConnectionResolver;
    let findManyResolverCalled;
    let countResolverCalled;

    beforeEach(() => {
      findManyResolverCalled = false;
      countResolverCalled = false;
      const mockedFindMany = userTC.getResolver('findMany').wrapResolve(next => resolveParams => {
        findManyResolverCalled = true;
        spyResolveParams = resolveParams;
        return next(resolveParams);
      });
      const mockedCount = userTC.getResolver('findMany').wrapResolve(next => resolveParams => {
        countResolverCalled = true;
        spyResolveParams = resolveParams;
        return next(resolveParams);
      });
      userTC.setResolver('mockedFindMany', mockedFindMany);
      userTC.setResolver('mockedCount', mockedCount);
      mockedConnectionResolver = prepareConnectionResolver(userTC, {
        countResolverName: 'mockedCount',
        findResolverName: 'mockedFindMany',
        sort: sortOptions,
      });
    });

    it('should pass to findMany args.sort', async () => {
      await mockedConnectionResolver.resolve({
        args: {
          sort: { name: 1 },
          first: 3,
        },
      });
      expect(spyResolveParams.args.sort.name).toBe(1);
    });

    it('should pass to findMany projection edges.node on top level', async () => {
      await mockedConnectionResolver.resolve({
        args: {},
        projection: {
          edges: {
            node: {
              name: true,
              age: true,
            },
          },
        },
      });
      expect(spyResolveParams.projection.name).toBe(true);
      expect(spyResolveParams.projection.age).toBe(true);
    });

    it('should pass to findMany custom projections to top level', async () => {
      await mockedConnectionResolver.resolve({
        args: {},
        projection: {
          score: { $meta: 'textScore' },
        },
      });
      expect(spyResolveParams.projection.score).toEqual({ $meta: 'textScore' });
    });

    it('should call count but not findMany when only count is projected', async () => {
      await mockedConnectionResolver.resolve({
        args: {},
        projection: {
          count: true,
        },
      });
      expect(countResolverCalled).toBe(true);
      expect(findManyResolverCalled).toBe(false);
    });

    it('should call count and findMany resolver when not only count is projected', async () => {
      await mockedConnectionResolver.resolve({
        args: {},
        projection: {
          count: true,
          edges: {
            node: {
              name: true,
              age: true,
            },
          },
        },
      });
      expect(countResolverCalled).toBe(true);
      expect(findManyResolverCalled).toBe(true);
    });

    it('should call findMany and not count when arbitrary top level fields are projected without count', async () => {
      await mockedConnectionResolver.resolve({
        args: {},
        projection: {
          name: true,
          age: true,
        },
      });
      expect(countResolverCalled).toBe(false);
      expect(findManyResolverCalled).toBe(true);
    });

    it('should call findMany and count when arbitrary top level fields are projected with count', async () => {
      await mockedConnectionResolver.resolve({
        args: {},
        projection: {
          count: true,
          name: true,
          age: true,
        },
      });
      expect(countResolverCalled).toBe(true);
      expect(findManyResolverCalled).toBe(true);
    });

    it('should call count and findMany resolver when last arg is used but not first arg', async () => {
      await mockedConnectionResolver.resolve({
        args: {
          last: 1,
        },
        projection: {
          edges: {
            node: {
              name: true,
              age: true,
            },
          },
        },
      });
      expect(countResolverCalled).toBe(true);
      expect(findManyResolverCalled).toBe(true);
    });

    it('should call findMany but not count resolver when first arg is used', async () => {
      await mockedConnectionResolver.resolve({
        args: { first: 1 },
        projection: {
          edges: {
            node: {
              name: true,
              age: true,
            },
          },
        },
      });
      expect(countResolverCalled).toBe(false);
      expect(findManyResolverCalled).toBe(true);
    });
  });

  describe('prepareRawQuery()', () => {
    const sortConfig = {
      value: { id: 1 },
      cursorFields: ['id'],
      beforeCursorQuery: (rawQuery, cursorData) => {
        rawQuery.before = cursorData;
        return rawQuery;
      },
      afterCursorQuery: (rawQuery, cursorData) => {
        rawQuery.after = cursorData;
        return rawQuery;
      },
    };

    it('should setup in resolveParams.rawQuery', () => {
      const rp: any = {
        args: { filter: { id: 123 } },
      };
      prepareRawQuery(rp, sortConfig);
      expect(rp.rawQuery).toEqual({});
    });

    it('should keep resolveParams.rawQuery if beforeCursorQuery/afterCursorQuery returns undefined', () => {
      const rawQuery = { a: 1 };
      const resolveParams = {
        args: {
          before: dataToCursor({ id1: 1 }),
          after: dataToCursor({ id2: 2 }),
        },
        rawQuery,
      };
      const dumbSortConfig = {
        value: { id: 1 },
        cursorFields: ['id'],
        beforeCursorQuery: () => {},
        afterCursorQuery: () => {},
      };
      prepareRawQuery(resolveParams, dumbSortConfig);
      expect(resolveParams.rawQuery).toBe(rawQuery);
    });

    it('should call afterCursorQuery if provided args.after', () => {
      const rp: any = {
        args: {
          after: dataToCursor({ id: 123 }),
          sort: sortConfig,
        },
      };
      prepareRawQuery(rp, sortConfig);
      expect(rp.rawQuery).toEqual({ after: { id: 123 } });
    });

    it('should call beforeCursorQuery if provided args.before', () => {
      const rp: any = {
        args: {
          before: dataToCursor({ id: 234 }),
          sort: sortConfig,
        },
      };
      prepareRawQuery(rp, sortConfig);
      expect(rp.rawQuery).toEqual({ before: { id: 234 } });
    });

    it('should call afterCursorQuery and beforeCursorQuery', () => {
      const rawQuery = { someKey: 1 };
      const resolveParams = {
        args: {
          before: dataToCursor({ id1: 1 }),
          after: dataToCursor({ id2: 2 }),
        },
        rawQuery,
      };
      prepareRawQuery(resolveParams, sortConfig);
      expect(resolveParams.rawQuery).toEqual({
        someKey: 1,
        before: { id1: 1 },
        after: { id2: 2 },
      });
    });
  });

  describe('preparePageInfo()', () => {
    const fiveEdges = [
      { cursor: 1, node: 1 },
      { cursor: 2, node: 2 },
      { cursor: 3, node: 3 },
      { cursor: 4, node: 4 },
      { cursor: 5, node: 5 },
    ];

    describe('"Relay Cursor Connections Specification (PageInfo)":', () => {
      describe('HasPreviousPage', () => {
        it('If last was not set (but first is present), return false.', () => {
          expect(preparePageInfo(fiveEdges, { first: 2 }, 5, 2).hasPreviousPage).toBe(false);
        });
        it('If last was not set (and first is empty), return true.', () => {
          expect(preparePageInfo(fiveEdges, {}, 5, 2).hasPreviousPage).toBe(true);
        });
        it('If last was not set (but after is present), return true.', () => {
          expect(preparePageInfo(fiveEdges, { after: 'abc' }, 5, 0).hasPreviousPage).toBe(true);
        });
        it('If edges contains more than last elements, return true.', () => {
          expect(preparePageInfo(fiveEdges, { last: 3 }, 3, 2).hasPreviousPage).toBe(true);
        });
        it('Return false', () => {
          expect(preparePageInfo(fiveEdges, { last: 5 }, 5, 0).hasPreviousPage).toBe(false);
        });
      });

      describe('HasNextPage', () => {
        it('If first was not set (and last is empty), return true.', () => {
          // By current Relay Cursor Connections Specification
          //   if `first` and `last` are empty `hasNextPage` should be false.
          // This rule is deviation from specification for better dev experience:
          //   when first and last args are empty
          //   we should check if exist more edges and provide correct `hasNextPage` value.
          expect(preparePageInfo(fiveEdges, {}, 4, 0).hasNextPage).toBe(true);
        });
        it('If first was not set (but last is present), return false.', () => {
          expect(preparePageInfo(fiveEdges, { last: 200 }, 4, 0).hasNextPage).toBe(false);
        });
        it('If edges contains more than first elements, return true.', () => {
          expect(preparePageInfo(fiveEdges, { first: 4 }, 4, 0).hasNextPage).toBe(true);
        });
        it('Return false', () => {
          expect(preparePageInfo(fiveEdges, { first: 5 }, 5, 0).hasNextPage).toBe(false);
        });
      });

      it('should return startCursor', () => {
        expect(preparePageInfo(fiveEdges, {}, 4, 0).startCursor).toBe(1);
        expect(preparePageInfo(fiveEdges, {}, 4, 2).startCursor).toBe(1);
      });

      it('should return endCursor', () => {
        expect(preparePageInfo(fiveEdges, {}, 4, 0).endCursor).toBe(4);
        expect(preparePageInfo(fiveEdges, {}, 20, 0).endCursor).toBe(5);
      });

      it('should return correct values for pageInfo if last is less first', async () => {
        const result = await connectionResolver.resolve({
          args: {
            sort: sortOptions.ID_ASC.value,
            first: 5,
            last: 3,
          },
        });
        expect(result.pageInfo.hasNextPage).toBe(true);
        expect(result.pageInfo.hasPreviousPage).toBe(true);
      });

      it('should return correct values for pageInfo if `last` toBe `first`', async () => {
        const result = await connectionResolver.resolve({
          args: {
            sort: sortOptions.ID_ASC.value,
            first: 5,
            last: 5,
          },
        });
        expect(result.pageInfo.hasNextPage).toBe(true);
        expect(result.pageInfo.hasPreviousPage).toBe(false);
      });

      it('should return correct values for pageInfo if set only `first`', async () => {
        const result = await connectionResolver.resolve({
          args: {
            sort: sortOptions.ID_ASC.value,
            first: 5,
          },
        });
        expect(result.pageInfo.hasNextPage).toBe(true);
        expect(result.pageInfo.hasPreviousPage).toBe(false);

        const result2 = await connectionResolver.resolve({
          args: {
            sort: sortOptions.ID_ASC.value,
            first: userList.length,
          },
        });
        expect(result2.pageInfo.hasNextPage).toBe(false);
        expect(result2.pageInfo.hasPreviousPage).toBe(false);
      });

      it('should return correct values for pageInfo if set only `last`', async () => {
        const result = await connectionResolver.resolve({
          args: {
            sort: sortOptions.ID_ASC.value,
            last: 5,
          },
        });
        expect(result.pageInfo.hasPreviousPage).toBe(true);
        expect(result.pageInfo.hasNextPage).toBe(false);

        const result2 = await connectionResolver.resolve({
          args: {
            sort: sortOptions.ID_ASC.value,
            last: userList.length,
          },
        });
        expect(result2.pageInfo.hasPreviousPage).toBe(false);
        expect(result2.pageInfo.hasNextPage).toBe(false);
      });
    });
  });

  describe('"Relay Cursor Connections Specification (Pagination algorithm)":', () => {
    describe('ApplyCursorsToEdges(allEdges, before, after):', () => {
      it('if `after` cursor is set, should return next record', async () => {
        const result = await connectionResolver.resolve({
          args: {
            after: dataToCursor({ id: 2 }),
            sort: sortOptions.ID_ASC.value,
            first: 1,
          },
        });
        expect(result.edges[0].node.id).toBe(3);
      });

      it('if `before` cursor is set, should return previous record', async () => {
        const result = await connectionResolver.resolve({
          args: {
            before: dataToCursor({ id: 2 }),
            sort: sortOptions.ID_ASC.value,
            first: 1,
          },
        });
        expect(result.edges[0].node.id).toBe(1);
      });

      it('if `before` and `after` cursors are set, should return between records', async () => {
        const result = await connectionResolver.resolve({
          args: {
            after: dataToCursor({ id: 2 }),
            before: dataToCursor({ id: 6 }),
            sort: sortOptions.ID_ASC.value,
            first: 10,
          },
        });
        expect(result.edges).toHaveLength(3);
        expect(result.edges[0].node.id).toBe(3);
        expect(result.edges[1].node.id).toBe(4);
        expect(result.edges[2].node.id).toBe(5);
      });

      it('should throw error if `first` is less than 0', async () => {
        const promise = connectionResolver.resolve({
          args: {
            first: -5,
            sort: sortOptions.ID_ASC.value,
          },
        });
        await expect(promise).rejects.toMatchSnapshot();
      });

      it('should slice edges to be length of `first`, if length is greater', async () => {
        const result = await connectionResolver.resolve({
          args: {
            sort: sortOptions.ID_ASC.value,
            first: 5,
          },
        });
        expect(result.edges).toHaveLength(5);
      });

      it('should throw error if `last` is less than 0', async () => {
        const promise = connectionResolver.resolve({
          args: {
            last: -5,
            sort: sortOptions.ID_ASC.value,
          },
        });
        await expect(promise).rejects.toMatchSnapshot();
      });

      it('should slice edges to be length of `last`', async () => {
        const result = await connectionResolver.resolve({
          args: {
            sort: sortOptions.ID_ASC.value,
            last: 3,
          },
        });
        expect(result.edges).toHaveLength(3);
      });

      it('should slice edges to be length of `last`, if `first` and `last` present', async () => {
        const result = await connectionResolver.resolve({
          args: {
            sort: sortOptions.ID_ASC.value,
            first: 5,
            last: 2,
          },
        });
        expect(result.edges).toHaveLength(2);
      });

      it('serve complex fetching with all connection args', async () => {
        const result = await connectionResolver.resolve({
          args: {
            sort: sortOptions.ID_ASC.value,
            after: dataToCursor({ id: 5 }),
            before: dataToCursor({ id: 13 }),
            first: 5,
            last: 3,
          },
          projection: {
            count: true,
            edges: {
              node: {
                name: true,
              },
            },
          },
        });
        expect(result.edges).toHaveLength(3);
        expect(result.edges[0].node.id).toBe(8);
        expect(result.edges[1].node.id).toBe(9);
        expect(result.edges[2].node.id).toBe(10);
        expect(result.count).toBe(15);
      });

      it('should correctly prepare cursor for before and after args', async () => {
        const result = await connectionResolver.resolve({
          args: {
            sort: sortOptions.ID_ASC.value,
            first: 3,
          },
        });
        expect(result.edges).toHaveLength(3);
        const { cursor } = result.edges[1];
        const prev = await connectionResolver.resolve({
          args: {
            sort: sortOptions.ID_ASC.value,
            first: 1,
            before: cursor,
          },
        });
        expect(prev.edges[0].node.id).toBe(result.edges[0].node.id);
        const conn = await connectionResolver.resolve({
          args: {
            sort: sortOptions.ID_ASC.value,
            first: 1,
            after: cursor,
          },
        });
        expect(conn.edges[0].node.id).toBe(result.edges[2].node.id);
      });
    });

    it('should return previous edges when querying with last and before', async () => {
      const result = await connectionResolver.resolve({
        args: {
          sort: sortOptions.ID_ASC.value,
          before: dataToCursor({ id: 4 }),
          last: 2,
        },
      });
      expect(result.edges).toHaveLength(2);
      expect(result.edges[0].node.id).toBe(2);
      expect(result.edges[1].node.id).toBe(3);
    });
  });

  describe('how works filter argument with resolve', () => {
    it('should add additional filtering', async () => {
      const result = await connectionResolver.resolve({
        args: {
          filter: {
            gender: 'm',
          },
          sort: sortOptions.ID_ASC.value,
          first: 100,
        },
        projection: {
          count: true,
          edges: {
            node: {
              name: true,
            },
          },
        },
      });
      expect(result.edges).toHaveLength(8);
      expect(result.edges[0].node).toEqual({
        id: 1,
        name: 'user01',
        age: 11,
        gender: 'm',
      });
      expect(result.edges[7].node).toEqual({
        id: 15,
        name: 'user15',
        age: 45,
        gender: 'm',
      });
      expect(result.count).toBe(8);
    });
  });

  describe('fallback logic (offset in cursor)', () => {
    it('if `after` cursor is set, should return next record', async () => {
      const result = await connectionResolver.resolve({
        args: {
          after: dataToCursor(1),
          sort: { name: 1 },
          first: 1,
        },
      });
      expect(result.edges[0].node.id).toBe(3);
    });

    it('if `before` cursor is set, should return previous record', async () => {
      const result = await connectionResolver.resolve({
        args: {
          before: dataToCursor(1),
          sort: { name: 1 },
          first: 1,
        },
      });
      expect(result.edges[0].node.id).toBe(1);
    });

    it('if `before` and `after` cursors are set, should return between records', async () => {
      const result = await connectionResolver.resolve({
        args: {
          after: dataToCursor(1),
          before: dataToCursor(5),
          sort: { name: 1 },
          first: 10,
        },
      });
      expect(result.edges).toHaveLength(3);
      expect(result.edges[0].node.id).toBe(3);
      expect(result.edges[1].node.id).toBe(4);
      expect(result.edges[2].node.id).toBe(5);
    });

    it('should throw error if `first` is less than 0', async () => {
      const promise = connectionResolver.resolve({
        args: {
          first: -5,
          sort: { name: 1 },
        },
      });
      await expect(promise).rejects.toMatchSnapshot();
    });

    it('should slice edges to be length of `first`, if length is greater', async () => {
      const result = await connectionResolver.resolve({
        args: {
          sort: { name: 1 },
          first: 5,
        },
      });
      expect(result.edges).toHaveLength(5);
    });

    it('should throw error if `last` is less than 0', async () => {
      const promise = connectionResolver.resolve({
        args: {
          last: -5,
          sort: { name: 1 },
        },
      });
      await expect(promise).rejects.toMatchSnapshot();
    });

    it('should slice edges to be length of `last`', async () => {
      const result = await connectionResolver.resolve({
        args: {
          sort: { name: 1 },
          last: 3,
        },
      });
      expect(result.edges).toHaveLength(3);
    });

    it('should slice edges to be length of `last`, if `first` and `last` present', async () => {
      const result = await connectionResolver.resolve({
        args: {
          sort: { name: 1 },
          first: 5,
          last: 2,
        },
      });
      expect(result.edges).toHaveLength(2);
    });

    it('serve complex fetching with all connection args', async () => {
      const result = await connectionResolver.resolve({
        args: {
          sort: { name: 1 },
          after: dataToCursor(4),
          before: dataToCursor(12),
          first: 5,
          last: 3,
        },
        projection: {
          count: true,
          edges: {
            node: {
              id: true,
            },
          },
        },
      });
      expect(result.edges).toHaveLength(3);
      expect(result.edges[0].node.id).toBe(8);
      expect(result.edges[1].node.id).toBe(9);
      expect(result.edges[2].node.id).toBe(10);
      expect(result.count).toBe(15);
    });

    it('should correctly prepare cursor for before and after args', async () => {
      const result = await connectionResolver.resolve({
        args: {
          sort: { name: 1 },
          first: 3,
        },
      });
      expect(result.edges).toHaveLength(3);
      const { cursor } = result.edges[1];
      const prev = await connectionResolver.resolve({
        args: {
          sort: { name: 1 },
          first: 1,
          before: cursor,
        },
      });
      expect(prev.edges[0].node.id).toBe(result.edges[0].node.id);
      const next = await connectionResolver.resolve({
        args: {
          sort: { name: 1 },
          first: 1,
          after: cursor,
        },
      });
      expect(next.edges[0].node.id).toBe(result.edges[2].node.id);
    });

    it('should reduce limit if reach cursor offset', async () => {
      const result = await connectionResolver.resolve({
        args: {
          sort: { name: 1 },
          before: dataToCursor(2),
          first: 5,
        },
      });
      expect(result.edges).toHaveLength(2);
      expect(result.edges[0].node.id).toBe(1);
      expect(result.edges[1].node.id).toBe(2);
    });
  });

  describe('default `first` argument if first/last are empty', () => {
    const defaultResolver = prepareConnectionResolver(userTC, {
      countResolverName: 'count',
      findResolverName: 'findMany',
      sort: sortOptions,
      defaultLimit: 5,
    });

    it('should use defaultLimit option', async () => {
      const data = await defaultResolver.resolve({ args: {} });
      expect(data.edges.length).toBe(5);
      expect(data.pageInfo.hasNextPage).toBe(true);
    });

    it('should use defaultLimit option with after option', async () => {
      const data = await defaultResolver.resolve({ args: { after: 'NA==' } });
      expect(data.edges.length).toBe(5);
      expect(data.pageInfo.hasNextPage).toBe(true);
      expect(data.pageInfo.hasPreviousPage).toBe(true);
    });
  });
});
