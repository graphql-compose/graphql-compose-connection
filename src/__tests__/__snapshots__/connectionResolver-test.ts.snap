// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`connectionResolver "Relay Cursor Connections Specification (Pagination algorithm)": ApplyCursorsToEdges(allEdges, before, after): should throw error if \`first\` is less than 0 1`] = `[Error: Argument \`first\` should be non-negative number.]`;

exports[`connectionResolver "Relay Cursor Connections Specification (Pagination algorithm)": ApplyCursorsToEdges(allEdges, before, after): should throw error if \`last\` is less than 0 1`] = `[Error: Argument \`last\` should be non-negative number.]`;

exports[`connectionResolver edges with data correctly handles filtering 1`] = `
Object {
  "count": 1,
  "edges": Array [
    Object {
      "cursor": "MA==",
      "id": 1,
      "node": Object {
        "age": 12,
        "gender": "m",
        "id": 2,
        "name": "user02",
      },
      "otherUserId": 2,
      "type": "likes",
      "userId": 1,
    },
  ],
  "pageInfo": Object {
    "endCursor": "MA==",
    "hasNextPage": false,
    "hasPreviousPage": false,
    "startCursor": "MA==",
  },
}
`;

exports[`connectionResolver edges with data correctly resolves with edges 1`] = `
Object {
  "count": 2,
  "edges": Array [
    Object {
      "cursor": "MA==",
      "id": 1,
      "node": Object {
        "age": 12,
        "gender": "m",
        "id": 2,
        "name": "user02",
      },
      "otherUserId": 2,
      "type": "likes",
      "userId": 1,
    },
    Object {
      "cursor": "MQ==",
      "id": 2,
      "node": Object {
        "age": 11,
        "gender": "m",
        "id": 1,
        "name": "user01",
      },
      "otherUserId": 1,
      "type": "dislikes",
      "userId": 2,
    },
  ],
  "pageInfo": Object {
    "endCursor": "MQ==",
    "hasNextPage": false,
    "hasPreviousPage": false,
    "startCursor": "MA==",
  },
}
`;

exports[`connectionResolver fallback logic (offset in cursor) should throw error if \`first\` is less than 0 1`] = `[Error: Argument \`first\` should be non-negative number.]`;

exports[`connectionResolver fallback logic (offset in cursor) should throw error if \`last\` is less than 0 1`] = `[Error: Argument \`last\` should be non-negative number.]`;
